jets/e/argon2.c

```diff
@@ -0,0 +1,145 @@
+    c3_y bytes_sat[wis];
+    u3r_bytes(0, wis, bytes_sat, sat);
+
+    c3_y outhash[out];
+    argon2_context context = {
+      outhash,             // output array, at least [digest length] in size
+      out,                 // digest length
+      bytes_dat,           // password array
+      wid,                 // password length
+      bytes_sat,           // salt array
+      wis,                 // salt length
+      bytes_key, wik,      // optional secret data
+      bytes_extra, wix,    // optional associated data
+      time_cost, mem_cost, threads, threads, // performance cost configuration
+      version,             // algorithm version
+      NULL, NULL,          // custom memory allocation / deallocation functions
+      ARGON2_DEFAULT_FLAGS // by default only internal memory is cleared
+    };
+
+    int argon_res;
+    switch(type)
+    {
+        case 100:   // %d
+      argon_res = argon2d_ctx(&context);
+      break;
+      //
+        case 105:   // %i
+      argon_res = argon2i_ctx(&context);
+      break;
+      //
+        case 117:   // %u
+      argon_res = argon2u_ctx(&context);
+      break;
+      //
+        case 25705: // %id
+      argon_res = argon2id_ctx(&context);
+      break;
+      //
+        default:
+      fprintf(stderr, "\nunjetted argon2 variant %i\n", type);
+      u3m_bail(c3__exit);
+    }
+
+    if(ARGON2_OK != argon_res) {
+      fprintf(stderr, "\nargon2 error: %s\n", argon2_error_message(argon_res));
+      u3m_bail(c3__exit);
+    }
+
+    u3z(key); u3z(extra); u3z(dat); u3z(sat);
+    return u3kc_rev(3, out, u3i_bytes(out, outhash));
+  }
+
+  u3_noun
+  u3we_argon2(u3_noun cor)
+  {
+    u3_noun // configuration params
+            out, type, version,
+            threads, mem_cost, time_cost,
+            wik, key, wix, extra,
+            // input params
+            wid, dat, wis, sat;
+
+    // the hoon code for argon2 takes configuration parameters,
+    // and then produces a gate. we jet that inner gate.
+    // this does mean that the config params have gotten buried
+    // pretty deep in the subject, so we have to use large
+    // tree addresses to get at them.
+    if ( (c3n == u3r_mean(cor, u3x_sam_4, &wid,
+                               u3x_sam_5, &dat,
+                               u3x_sam_6, &wis,
+                               u3x_sam_7, &sat,
+                             //
+                               1020,      &out,
+                               2042,      &type,
+                               4086,      &version,
+                             //
+                               8174,      &threads,
+                               16350,     &mem_cost,
+                               32702,     &time_cost,
+                             //
+                               130812,    &wik,
+                               130813,    &key,
+                               130814,    &wix,
+                               130815,    &extra, 0)) ||
```

Comment:
I'm a little uncomfortable about pulling out things by specific deep axis; it feels really fragile. That might just be me though.

Reply:
Yes, if the interface for the jetted arm changes in any way, this will need to be changed. But I think that's the case for all jets? Is there an alternative you'd expect/propose?

(Github brought you up as a suggestion, so I just tagged you. Feel free to defer to defer.)